xapian-glib/xapian-term-iterator.cc

```diff
@@ -0,0 +1,317 @@
+
+  iter = static_cast<XapianTermIterator *> (g_object_new (XAPIAN_TYPE_TERM_ITERATOR, NULL));
+  priv = XAPIAN_TERM_ITERATOR_GET_PRIVATE (iter);
+  priv->data = new IteratorData (it);
+
+  return iter;
+}
+
+/**
+ * xapian_term_iterator_is_valid:
+ * @iter: a #XapianTermIterator
+ *
+ * Checks whether @iter is valid.
+ *
+ * Returns: %TRUE if the iterator is valid
+ *
+ * Since: 1.4
+ */
+gboolean
+xapian_term_iterator_is_valid (XapianTermIterator *iter)
+{
+  g_return_val_if_fail (XAPIAN_IS_TERM_ITERATOR (iter), FALSE);
+
+  return XAPIAN_TERM_ITERATOR_GET_PRIVATE (iter)->data != NULL;
+}
+
+/**
+ * xapian_term_iterator_next:
+ * @iter: a #XapianTermIterator
+ *
+ * Advances the @iter forward.
+ *
+ * You should use the return value to check if @iter is still
+ * valid after being advanced, before using the #XapianTermIterator
+ * API.
+ *
+ * Returns: %TRUE if the iterator was advanced, and %FALSE otherwise
+ *
+ * Since: 1.4
+ */
+gboolean
+xapian_term_iterator_next (XapianTermIterator *iter)
+{
+  g_return_val_if_fail (XAPIAN_IS_TERM_ITERATOR (iter), FALSE);
+
+  XapianTermIteratorPrivate *priv = XAPIAN_TERM_ITERATOR_GET_PRIVATE (iter);
+
+  if (priv->data == NULL)
+    return FALSE;
+
+  return priv->data->next ();
+}
+
+/**
+ * xapian_term_iterator_get_term_name:
+ * @iter: a #XapianTermIterator
+ *
+ * Retrieves the term name of the current item pointed by @iter.
+ *
+ * Returns: (transfer full): the term name
+ *
+ * Since: 1.4
+ */
+char *
+xapian_term_iterator_get_term_name (XapianTermIterator *iter)
+{
+  g_return_val_if_fail (XAPIAN_IS_TERM_ITERATOR (iter), 0);
+
+  XapianTermIteratorPrivate *priv = XAPIAN_TERM_ITERATOR_GET_PRIVATE (iter);
+
+  if (priv->data == NULL)
+    return NULL;
+
+  return priv->data->getTermName ();
+}
+
+/**
+ * xapian_term_iterator_get_wdf:
+ * @iter: a #XapianTermIterator
+ * @error: return location for a #GError
+ *
+ * Retrieves the WDF of the current item pointed by @iter.
+ *
+ * Returns: the WDF
+ *
+ * Since: 1.4
+ */
+unsigned int
+xapian_term_iterator_get_wdf (XapianTermIterator *iter,
+                              GError            **error)
+{
+  g_return_val_if_fail (XAPIAN_IS_TERM_ITERATOR (iter), 0);
+
+  XapianTermIteratorPrivate *priv = XAPIAN_TERM_ITERATOR_GET_PRIVATE (iter);
+
+  if (priv->data == NULL)
+    return 0;
+
+  try
+    {
+      return priv->data->getWdf ();
+    }
+  catch (const Xapian::Error &err)
+    {
+      GError *internal_error = NULL;
+
+      xapian_error_to_gerror (err, &internal_error);
+      g_propagate_error (error, internal_error);
+
+      return 0;
+    }
+}
+
+/**
+ * xapian_term_iterator_get_term_freq:
+ * @iter: a #XapianTermIterator
+ * @error: return location for a #GError
+ *
+ * Retrieves the term frequency of the current item pointed by @iter.
+ *
+ * Returns: the term frequency
+ *
+ * Since: 1.4
+ */
+unsigned int
+xapian_term_iterator_get_term_freq (XapianTermIterator *iter,
+                                    GError            **error)
```

Comment:
You shouldn't be able to get 0 here (because any term returned should exist in the database, which means a non-zero term frequency).

To return `-1` the return type would need to be changed, but then it wouldn't match the existing `xapian_database_get_term_freq` which returns `unsigned int` (ironically, that one definitely can return `0` - it will if asked for the term frequency of a term which isn't present in any document in the database).

The main error case here is that not all `TermIterator` objects actually support this method and those that don't will throw `Xapian::InvalidOperationError`.  Perhaps it'd be simpler to just document that those return 0 and handle such exceptions that way instead of via `GError`.  You can also get other exceptions like `Xapian::DatabaseError` here, but you can from many Xapian methods, and xapian-glib doesn't generally try to handle that.

FWIW, even `unsigned int` is kind of wrong as it should really be an unsigned integer type equivalent to `Xapian::doccount` - it'd be better really to have an equivalent C `typedef` (say `xapian_doccount`) here, though not sure if it's worth retrofitting.  If you discount platforms with 16 bit `int` (which I suspect building on glib inherently does) then in practical terms this just means xapian-glib users can't take full advantage of xapian-core configured with `--enable-64bit-docid`.